sdk/common/src/main/java/io/opentelemetry/sdk/entities/SdkEntityProvider.java

```diff
@@ -0,0 +1,134 @@
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class SdkEntityProvider implements EntityProvider {
+
+  private final Object lock = new Object();
+  private final AtomicReference<Resource> resource = new AtomicReference<>(Resource.empty());
+
+  @GuardedBy("lock")
+  private final LinkedHashMap<String, Entity> entities = new LinkedHashMap<>();
+
+  @GuardedBy("lock")
+  private final List<EntityListener> listeners = new ArrayList<>();
+
+  public SdkEntityProvider() {
+    this(Collections.emptyList());
+  }
+
+  public SdkEntityProvider(List<Entity> initialEntities) {
+    for (Entity entity : initialEntities) {
+      entities.put(entity.getId(), entity);
+    }
+    rebuildResource();
+  }
+
+  public static EntityProvider getDefault() {
+    Resource resource = Resource.getDefault();
+    Entity entity = Entity.create(resource);
+    return new SdkEntityProvider(singletonList(entity));
+  }
+
+  public static EntityProvider empty() {
+    return new SdkEntityProvider(singletonList(Entity.create(Resource.empty())));
+  }
+
+  @Override
+  public Resource getResource() {
+    Resource result = resource.get();
+    return result == null ? Resource.empty() : result;
+  }
+
+  @Override
+  public void addListener(EntityListener listener) {
+    synchronized (lock) {
+      listeners.add(listener);
+    }
+  }
+
+  @Override
+  public void addEntity(String id, String name, Attributes attributes) {
+    Entity entity = Entity.create(id, name, attributes);
+    List<EntityListener> listeners;
+    Resource resource;
+    synchronized (lock) {
+      entities.remove(id);
+      entities.put(id, entity);
+
+      resource = rebuildResource();
+      listeners = new ArrayList<>(this.listeners);
+    }
+    for (EntityListener listener : listeners) {
+      listener.onEntityState(entity, resource);
+    }
+  }
+
+  @Override
+  public void updateEntity(String id, Attributes attributes) {
+    List<EntityListener> listeners;
+    Resource resource;
+    Entity updatedEntity;
+    synchronized (lock) {
+      Entity entity = entities.get(id);
+      if (entity == null) {
+        return;
+      }
+      updatedEntity = entity.withAttributes(attributes);
+      entities.put(id, updatedEntity);
+      resource = rebuildResource();
+      listeners = new ArrayList<>(this.listeners);
+    }
+    for (EntityListener listener : listeners) {
+      listener.onEntityState(updatedEntity, resource);
+    }
+  }
+
+  @Override
+  public void deleteEntity(String id) {
+    Entity removedEntity;
+    Resource resource;
+    synchronized (lock) {
+      removedEntity = entities.remove(id);
+      if (removedEntity == null) {
+        return;
+      }
+      resource = rebuildResource();
+    }
+    for (EntityListener listener : listeners) {
+      listener.onEntityDelete(removedEntity, resource);
+    }
+  }
+
+  private Resource rebuildResource() {
+    Resource newResource = doRebuildResource();
+    resource.set(newResource);
```

Comment:
nit, you can use `lazySet` here to avoid flushing your local cache.  The details on this are quite obtuse over time, but here's the best explanation (that I believe is still valid) for why: https://psy-lob-saw.blogspot.com/2012/12/atomiclazyset-is-performance-win-for.html

I believe you're abiding by single-writer principle, and this could help avoiding flushing cache on write until the time the rest of the SDK requests access to the resource.